drivers/epd_bw_spi/epd_bw_spi_disp_dev.c

```diff
@@ -0,0 +1,128 @@
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_epd_bw_spi
+ * @{
+ *
+ * @file
+ * @brief       Driver adaption to disp_dev generic interface
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @}
+ */
+
+#include <assert.h>
+#include <stdint.h>
+
+#include "epd_bw_spi.h"
+#include "epd_bw_spi_disp_dev.h"
+
+#ifndef EPD_BW_SPI_DISP_COLOR_DEPTH
+#define EPD_BW_SPI_DISP_COLOR_DEPTH (1U)
+#endif
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static void _epd_bw_spi_map_full(epd_bw_spi_t *dev,
+                                 uint16_t x1, uint16_t x2,
+                                 uint16_t y1, uint16_t y2,
+                                 const uint16_t *color)
+{
+    uint16_t size = ((x2 - x1) >> 3) * (y2 - y1);
+
+    epd_bw_spi_init_full(dev);
+    epd_bw_spi_activate(dev);
+    epd_bw_spi_set_area(dev, x1, x2, y1, y2);
+    epd_bw_spi_write_buffer(dev,  (uint8_t *)color, size);
+    epd_bw_spi_update_full(dev);
+}
+
+static void _epd_bw_spi_map_part(epd_bw_spi_t *dev,
+                                 uint16_t x1, uint16_t x2,
+                                 uint16_t y1, uint16_t y2,
+                                 const uint16_t *color)
+{
+    uint16_t size = ((x2 - x1) >> 3) * (y2 - y1);
+
+    DEBUG("[epd_bw_spi] map part (%d,%d)-(%d,%d) (%u bytes)\n",
+          x1, y1, x2, y2, size);
+    epd_bw_spi_init_part(dev);
+    epd_bw_spi_activate(dev);
+    epd_bw_spi_set_area(dev, x1, x2, y1, y2);
+    epd_bw_spi_write_buffer(dev, (uint8_t *)color, size);
+    epd_bw_spi_update_part(dev);
+}
+
+static void _epd_bw_spi_map(disp_dev_t *disp_dev, uint16_t x1, uint16_t x2,
+                            uint16_t y1, uint16_t y2, const uint16_t *color)
+{
+    epd_bw_spi_t *dev = (epd_bw_spi_t *)disp_dev;
+
+    /* Wake the device from sleep */
+    epd_bw_spi_wake(dev);
+
+    /* Update the display */
+    if (dev->refresh_partial) {
+        _epd_bw_spi_map_part(dev, x1, x2, y1, y2, color);
+    }
+    else {
+        _epd_bw_spi_map_full(dev, x1, x2, y1, y2, color);
+    }
+
+    /* Put the display to sleep */
+    epd_bw_spi_deactivate(dev);
+    if (dev->params.rst_pin != GPIO_UNDEF) {
+        epd_bw_spi_sleep(dev);
+    }
+}
+
+static uint16_t _epd_bw_spi_height(disp_dev_t *disp_dev)
+{
+    epd_bw_spi_t *dev = (epd_bw_spi_t *)disp_dev;
+
+    assert(dev);
+
+    return dev->size_y;
+}
+
+static uint16_t _epd_bw_spi_width(disp_dev_t *disp_dev)
+{
+    const epd_bw_spi_t *dev = (epd_bw_spi_t *)disp_dev;
+
+    assert(dev);
+
+    return dev->size_x;
+}
+
+static uint8_t _epd_bw_spi_color_depth(disp_dev_t *disp_dev)
+{
+    (void)disp_dev;
+    return EPD_BW_SPI_DISP_COLOR_DEPTH;
+}
+
+static void _epd_bw_spi_set_invert(disp_dev_t *disp_dev, bool invert)
+{
+    (void)disp_dev;
+    (void)invert;
+    assert(false);
+}
+
+void epd_bw_spi_set_partial(disp_dev_t *disp_dev, bool partial)
+{
+    ((epd_bw_spi_t *)disp_dev)->refresh_partial = partial;
+}
```

Comment:
Always using a partial refresh will lead to horrible ghosting, so I don't think it's a good idea to always perform partial refreshes. On the other hand, always doing full refreshes results in a suboptimal display experience and increased power use.

Maybe instead of a `refresh_partial` flag it should be a `refresh_auto` flag that uses an internal counter to perform a full refresh once every `n` refreshes?

Reply:
I don't have a strong opinion on this. If this logic is implemented within the disp_dev maps function, then why not. The value of `n` could be an overridable define.